### PR TITLE
Adding color chooser preferences

### PR DESCRIPTION
**Pull request by MarcelS4**
We need the ability to select which color choosers are available/allowed
I have therefore started adding a Color chooser preferences page to the
Window builder preferces

Removing unnecessary preference options

The preference that allows the user to order the tabs in the color
property editor is deemed unnecessary. It has therefore been removed.

Updating ColorChooser preference to disregard changes that were made and
not applied. I.e. setting the checkboxes to the previously saved state

Removing unnecessary constants from the IColorChooserPreferenceConstants

Updating ModelMessages to contain the new string values used on the
preferences

Updating IColorChooserPreferenceConstants to obtain the names of the
various options from ModelMessages instead of hardcoding the string
values

### CODE REVIEW

**Comment by merks** [comment]
This appears to add a new jar.  What is the origin of this JAR? What license and copyrights are applicable to it?

**Comment by MarcelS4** [comment]
Hi @merks  this jar was added when I created the layout for the preference page, I did not add it manually but after a quick search the origin and license can be found at 
https://mvnrepository.com/artifact/org.eclipse.wb.rcp/swing2swt/1.6.0

**Comment by merks** [comment]
Hopefully someone who understand the design intent will comment but it doesn't seem like a good thing to have copies of binaries in the source repository...

**Comment by MarcelS4** [comment]
Where are the binaries usually placed? I'd be happy to make changes as needed

**Comment by MarcelS4** [comment]
UPDATE: I've tried referencing the swing2swt jar that can be found in the project org.eclipse.wb.rcp.swing2swt, however this produces an error that it is unable to create the BorderLayout

**Comment by vogella** [comment]
I would also suggest to remove the jar from the commit. Maybe add it to the .gitignore?

**Comment by MarcelS4** [comment]
I can remove it from the commit, but as I mentioned, I tried removing the jar from the org.eclipse.wb.swing project. After I did this I referenced the swing2swt.jar from the org.eclipse.wb.rcp.swing2swt project but this produces errors when trying to load the preference page.

**Comment by vogella** [comment]
I think the jar is created by the build process so it is not IMHO not necessary to commit it.

**Comment by MarcelS4** [comment]
My mistake. Thank you @vogella. I'll add it to the gitignore

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java [hunk 1]
```diff
@@ -0,0 +1,149 @@
+package org.eclipse.wb.internal.swing.model.property.editor.color;
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java:1]: Please add your copyright here. Look at other classes.

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/ColorChooserPreferences.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
+package org.eclipse.wb.internal.swing.preferences.colorchooser;
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/ColorChooserPreferences.java:1]: Please add the license

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/ColorChooserPreferences.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+import org.osgi.service.prefs.BackingStoreException;
+import org.osgi.service.prefs.Preferences;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import swing2swt.layout.BorderLayout;
+
+public class ColorChooserPreferences extends PreferencePage implements IWorkbenchPreferencePage {
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/ColorChooserPreferences.java:26]: Please add Javadoc.

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/ColorChooserPreferences.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+        ColorPropertyEditor.reloadColorDialog();
+      }
+    } catch (BackingStoreException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * This method is called on the "Apply and Close" button
+   */
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/ColorChooserPreferences.java:159]: If your method does something else than it describes in the superclass then add JavaDoc preferably with the @inheritdoc tag. Otherwise, please remove the javadoc from inherited methods.

**Review by wimjongman — CHANGES_REQUESTED** [review]
Thanks for adding this to WB. It is very much appreciated. Please take a look at my review.

**Comment by MarcelS4** [comment]
@wimjongman thank you so much for the feedback. I've made the changes you requested. I was a bit unsure about the copyright so I followed what was done on the other pages. Please let me know if there is anything I need to change

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
+/*******************************************************************************
+ * Copyright (c) DSA.
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java:2]: Normally it has a first and last change year date. I'm not sure if it is required:

Copyright (c) 2021, 2021 DSA

Also, who is DSA? I could not find it on google. you have to uniquely identify the copyright holder. Maybe at a web address:

DSA Automobile Company (https://dsa.com)

Please also note that your company should agree that they put this in open source:

https://www.eclipse.org/legal/legalfaq.php#h.l3bvwrmty0iz

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/IColorChooserPreferenceConstants.java [hunk 6]
```diff
@@ -0,0 +1,20 @@
+package org.eclipse.wb.internal.swing.preferences.colorchooser;
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/IColorChooserPreferenceConstants.java:1]: Copyright must be added to all new files!

**Review by wimjongman — CHANGES_REQUESTED** [review]

**Comment by wimjongman** [comment]
Thanks, Marcel! I made some more comments. Sorry :)

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java (hunk 5 above)
Comment (MarcelS4) [reply org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java:2]: Thank you for the feedback. :) I've spoken to my team lead at DSA about the public license and as far as is my understanding they have agreed to put that code in open source. I'll add their web address to the licenses and just ask my team lead to confirm with whomever he has to confirm with before making another commit.

When you say "first and last change year date" you're referring to the code changes right?

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/IColorChooserPreferenceConstants.java (hunk 6 above)
Comment (MarcelS4) [reply org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/preferences/colorchooser/IColorChooserPreferenceConstants.java:1]: I understand I will do so. Thank you. :)

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java (hunk 5 above)
Comment (wimjongman) [reply org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java]: Yes, changes in that source.

File: org.eclipse.wb.swing/META-INF/MANIFEST.MF [hunk 7]
```diff
@@ -3,7 +3,8 @@ Bundle-ManifestVersion: 2
 Bundle-Name: %pluginName
 Bundle-SymbolicName: org.eclipse.wb.swing;singleton:=true
 Bundle-Version: 1.9.1.qualifier
-Bundle-ClassPath: .
+Bundle-ClassPath: .,
+ swing2swt.jar
```
Comment (wimjongman) [inline org.eclipse.wb.swing/META-INF/MANIFEST.MF:7]: Since that library is not generated anymore, it should not be added to gitignore. Please remote the entry.

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/ColorPropertyEditor.java [hunk 8]
```diff
@@ -7,6 +7,7 @@
  *
  * Contributors:
  *    Google, Inc. - initial API and implementation
+ *    Marcel du Preez - Color preferences integration and System color warning
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/ColorPropertyEditor.java:10]: 👍 Please also change the copyright date in every source you change.

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
+/*******************************************************************************
+ * Copyright (c) 2021, 2021 DSA Daten- und Systemtechnik GmbH. (https://www.dsa.de)
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java:2]: 👍 excellent.

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java (hunk 5 above)
Comment (MarcelS4) [reply org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java]: Okay thank you. Then I believe the changes I made should be right? Sorry for all the confusion. Im new to WB and git

**Review by wimjongman — COMMENTED** [review]
Almost there, Marcel.

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java (hunk 5 above)
Comment (wimjongman) [reply org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java]: No problem, Marcel. Thank you for contributing. You can ask me anything. Please ping me if I am not quick enough to respond ('pinging': reply on the issue with "ping".)

**Comment by MarcelS4** [comment]
Thank you @wimjongman

**Comment by wimjongman** [comment]
> 
> 
> Thank you @wimjongman for you patience and guidance. It's greatly appreciated

Sure! Please note that you should not force-push. Just keep committing and pushing. We can squash everything before we merge it back.

If you force-push GitHub loses track of comments in the review.

File: org.eclipse.wb.swing/.gitignore [hunk 10]
```diff
@@ -1 +1,2 @@
 /bin
+/swing2swt.jar
```
Comment (wimjongman) [inline org.eclipse.wb.swing/.gitignore:2]: Please remove. It is no longer needed.

Comment (MarcelS4) [reply org.eclipse.wb.swing/.gitignore]: Apologies, I thought I had removed it. I have done so again. Hoping it's right this time :)

**Comment by MarcelS4** [comment]
Thank you. I'll remember for the future

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/ColorPropertyEditor.java (hunk 8 above)
Comment (wimjongman) [reply org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/ColorPropertyEditor.java:10]: You did not add "and others". Please do it next time.

**Review by wimjongman — APPROVED** [review]

File: org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java [hunk 11]
```diff
@@ -315,6 +313,7 @@ private static String getSource(GenericProperty property, ColorInfo colorInfo) t
     }
     IPreferenceStore preferences =
         property.getJavaInfo().getDescription().getToolkit().getPreferences();
+    preferences.addPropertyChangeListener(null);
```
Comment (wimjongman) [inline org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java:316]: > preferences.addPropertyChangeListener(null);

What is the intended use of this? Are you using it for its side-effects? 

image

File: org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+        colorTextHue.setText("" + String.format("%.2f", currentColor.getHSB()[0]));
+        colorTextSaturation.setText("" + String.format("%.2f", currentColor.getHSB()[1]));
+        colorTextbrightness.setText("" + String.format("%.2f", currentColor.getHSB()[2]));
+      }
+
+      public void widgetDefaultSelected(SelectionEvent e) {
+        // TODO Auto-generated method stub
+      }
+    });
+  }
```
Comment (wimjongman) [inline org.eclipse.wb.swing/src/org/eclipse/wb/internal/swing/model/property/editor/color/CustomColorPickerComposite.java:93]: To avoid empty methods, use SelectionAdapter instead of SelectionListener.

you can also use lambda 

    colorSelector.getButton().addListener(SWT.Selection, e->{magic here});

**Comment by wimjongman** [comment]
Marcel, we also need a linked issue. On the issue can you please describe what you have done also with pictures?

**Comment by vogella** [comment]
Would be nice to make the unit tests work again before merging features to avoid breaks.  See #42

File: org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java [hunk 13]
```diff
@@ -217,10 +217,8 @@ private static String getColorFieldName(Expression idExpression) throws Exceptio
       }
     }
     // no field, unexpected
-    throw new IllegalArgumentException(MessageFormat.format(
-        ModelMessages.ColorPropertyEditor_wrongSwtColor,
-        idExpression,
-        id));
+    throw new IllegalArgumentException(
```
Comment (vogella) [inline org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java:220]: Format changes and real changes should not be mixed. I like this reformat, please provide separate PR for this.

File: org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java [hunk 14]
```diff
@@ -176,10 +177,9 @@ private String getText(Property property) throws Exception {
           Object keyArgument = invocation.arguments().get(0);
           if (keyArgument instanceof QualifiedName) {
             GenericProperty genericProperty = (GenericProperty) property;
-            ResourceRegistryInfo registry =
-                RegistryContainerInfo.getRegistry(
-                    genericProperty.getJavaInfo().getRootJava(),
-                    invocation.getExpression());
+            ResourceRegistryInfo registry = RegistryContainerInfo.getRegistry(
```
Comment (vogella) [inline org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java:180]: Format changes and real changes should not be mixed. I like this reformat, please provide separate PR for this.

File: org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java [hunk 15]
```diff
@@ -453,11 +451,10 @@ public RegistryColorsPage(Composite parent,
           }
         }
         // create grid composite
-        ColorsGridComposite colorsGrid =
-            createColorsGroup(
-                this,
-                registryInfo.getVariableSupport().getTitle(),
-                infos.toArray(new ColorInfo[infos.size()]));
+        ColorsGridComposite colorsGrid = createColorsGroup(
```
Comment (vogella) [inline org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java:454]: Format changes and real changes should not be mixed. please provide separate PR for this.

**Review by vogella — COMMENTED** [review]
Would you update the PR by squashing both commits into one?

Also looks to me that you mix format changes and real changes. Can you separate the format changes from the real changes via a separate pull request?


Also did you try to run the wb.tests before and after this change from your IDE? Any new errorsw with the change?

File: org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java (hunk 15 above)
Comment (wimjongman) [reply org.eclipse.wb.swt/src/org/eclipse/wb/internal/swt/model/property/editor/color/ColorPropertyEditor.java:454]: Lars, maybe we can allow this for now and ask Marcel to take this into account for later pushes?

**Comment by vogella** [comment]
@wimjongman what is the status here? Do you still plan to review / finish this review or shall I take over?

**Comment by wimjongman** [comment]
Yes, please take over. Lars, why are you asking to squash commits and force push? When this is done, the inline review is removed.

https://github.com/eclipse/windowbuilder/pull/91#issuecomment-970265327

**Comment by vogella** [comment]
Please rebase onto master

**Review by vogella — COMMENTED** [review]
faeb9aa05f3de6c9b79738a7aad8cfb11fded20f commits seems unintended. Please remove and force push an udpate.

File: org.eclipse.wb.swing/.classpath [hunk 16]
```diff
@@ -1,7 +1,7 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<classpath>
-	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/J2SE-1.5"/>
-	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
-	<classpathentry kind="src" path="src"/>
-	<classpathentry kind="output" path="bin"/>
-</classpath>
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vogella) [inline org.eclipse.wb.swing/.classpath:1]: This looks unrelated, please update wb.swing to Java 11 (via the MANIFEST.MF) and update the classpath

Please send a new PR for this and remove this change from this PR.

image

**Review by vogella — CHANGES_REQUESTED** [review]

File: org.eclipse.wb.swing/.classpath (hunk 16 above)
Comment (MarcelS4) [reply org.eclipse.wb.swing/.classpath:1]: @vogella I've made the changes you requested

Comment (MarcelS4) [reply org.eclipse.wb.swing/.classpath:1]: @vogella  requested changes applied

**Comment by vogella** [comment]
@MarcelS4 can you put the Java 11 update into a separate PR and rebase this one onto master once these are in?

**Comment by MarcelS4** [comment]
Rebase onto latest master complete

**Comment by vogella** [comment]
@MarcelS4 still two commits, you can check this on https://github.com/eclipse/windowbuilder/pull/91/commits

**Comment by MarcelS4** [comment]
> @MarcelS4 still two commits, you can check this on https://github.com/eclipse/windowbuilder/pull/91/commits

Apologies Lars, I rebased and pushed them separately

**Comment by vogella** [comment]
Trying to figure out why I do not see the test results in the pull requests from @MarcelS4 similar to https://github.com/eclipse/windowbuilder/pull/230#issuecomment-1128583806

**Comment by MarcelS4** [comment]
> Trying to figure out why I do not see the test results in the pull requests from @MarcelS4 similar to [#230 (comment)](https://github.com/eclipse/windowbuilder/pull/230#issuecomment-1128583806)

Anything I can do to assist?

**Comment by wimjongman** [comment]
Marcel has to rebase first to get the new actions.

**Comment by MarcelS4** [comment]
> Trying to figure out why I do not see the test results in the pull requests from @MarcelS4 similar to [#230 (comment)](https://github.com/eclipse/windowbuilder/pull/230#issuecomment-1128583806)

@vogella I've rebased onto latest master as @wimjongman suggested. Please let me know if this resolves the issue. I'll start rebasing the other PR's as well

**Comment by vogella** [comment]
As discussed today @MarcelS4 I will have more confidence in reviewing any PR in windowsbuilder if we have more unit tests running, currently have ~3600 more working tests from the core suite, with only 44 left which are failing.

image

Once this is working I may ask you to help with the remaining test suites to ensure WB stays stable. :-o

**Comment by vogella** [comment]
Tests are running fine with this change. Now manually testing.

**Comment by MarcelS4** [comment]
@vogella changes were made to create a Preference page for both SWT and Swing.

**Comment by MarcelS4** [comment]
Ignore last comment. 

I found a mistake that needs rectifying.

**Comment by MarcelS4** [comment]
Closes #95
